### PR TITLE
register and expose services on kernel

### PR DESCRIPTION
**Pull request by colombod**

### CODE REVIEW

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@ public abstract class KernelBase : IKernel
         private readonly CompositeDisposable _disposables;
         private readonly SubmissionSplitter _submissionSplitter = new SubmissionSplitter();
         private readonly ConcurrentQueue<IKernelCommand> _deferredCommands = new ConcurrentQueue<IKernelCommand>();
+        private readonly ConcurrentDictionary<Type, object> _propertyBag = new ConcurrentDictionary<Type, object>();
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelBase.cs]: Nit: `_properties`

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 2]
```diff
@@ -289,5 +292,15 @@ public void RegisterForDisposal(IDisposable disposable)
         public void Dispose() => _disposables.Dispose();
 
         string IKernel.Name => Name;
+
+        public void SetProperty<T>(T service) where T : class
```
[the remark this answers is not in the document]
Comment (colombod) [reply Microsoft.DotNet.Interactive/KernelBase.cs]: I think is better to fail silently on Get but fail on failure to set

File: Microsoft.DotNet.Interactive/KernelBase.cs (hunk 1 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/KernelBase.cs]: done

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 3]
```diff
@@ -289,5 +292,18 @@ public void RegisterForDisposal(IDisposable disposable)
         public void Dispose() => _disposables.Dispose();
 
         string IKernel.Name => Name;
+
+        public void SetProperty<T>(T property) where T : class
+        {
+            if (!_propertyBag.TryAdd(typeof(T), property))
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelBase.cs]: If you use `Add` you can just let the default exception be thrown.

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 4]
```diff
@@ -289,5 +292,18 @@ public void RegisterForDisposal(IDisposable disposable)
...
+            {
+                throw new InvalidOperationException($"Cannot add property with key {typeof(T)}.");
+            }
+        }
+
+        public T GetProperty<T>() where T : class
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelBase.cs]: This should throw if the property is not found.

File: Microsoft.DotNet.Interactive/KernelBase.cs (hunk 3 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/KernelBase.cs]: I don’t think there is Add on concurrent dictionary

File: Microsoft.DotNet.Interactive/KernelBase.cs (hunk 4 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/KernelBase.cs]: ok, makes sense

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 5]
```diff
@@ -187,6 +188,8 @@ private async Task ExecuteCommand(KernelOperation operation)
         private readonly ConcurrentQueue<KernelOperation> _commandQueue =
             new ConcurrentQueue<KernelOperation>();
 
+  
+
```
Comment (daxian-dbw) [inline Microsoft.DotNet.Interactive/KernelBase.cs]: nit: extra new lines that you may want to remove?

File: Microsoft.DotNet.Interactive/KernelBase.cs [hunk 6]
```diff
@@ -289,5 +292,20 @@ public void RegisterForDisposal(IDisposable disposable)
...
+            {
+                throw new InvalidOperationException($"Cannot add property with key {typeof(T)}.");
+            }
+        }
+
+        public T GetProperty<T>() where T : class
+        {
+            return _propertyBag.TryGetValue(typeof(T), out var property) 
+                ? property as T 
+                : throw new InvalidOperationException($"Cannot find property {typeof(T)}.");
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelBase.cs]: Maybe `KeyNotFoundException`?

**Pull request merged**